### ISSUE TITLE
Processing images for publishing

### ISSUE THREAD

**Issue by mkroman** [getzola/zola#838]
Much like how Zola has the ability to resize images, it would be nice if it was able to process images and wipe any sensitive EXIF data, such as geo-coordinates.

This shouldn't be too hard to implement as it can be done similar to [this example](https://www.getzola.org/documentation/content/image-processing/#using-resize-image-in-markdown-via-shortcodes) just by using a `process_image` function.

There's one caveat to this, though! Stripping the EXIF data can mean losing orientation information as a lot of modern smartphones simply save the orientation as metadata rather than rotate the image.

Another non-trivial concern is that people like me who publish the [blog source](https://github.com/mkroman/blog) might check in an image with the same sensitive data.

One solution to this could be recommending a default `.gitignore` with a directory specifically for images like this, and `process_image` could work even when the source image isn't found, and it will skip the processing if the determined output image already exists in the repo.
It's not a perfect solution, though.

**Comment by Keats** [comment]
> There's one caveat to this, though! Stripping the EXIF data can mean losing orientation information as a lot of modern smartphones simply save the orientation as metadata rather than rotate the image.

That would have been too simple :(

There are already 2 images functions (`resize` and `get_image_metadata`) so I'm not against adding one more but it's becoming a bit hard to understand how they work together: can I combine them? Considering they would both create an image that seems like it would inflate the size of the repo/site. A function to *read* EXIF data could be super useful too, at least to get lat/long/date.

> One solution to this could be recommending a default .gitignore with a directory specifically for images like this, and process_image could work even when the source image isn't found, and it will skip the processing if the determined output image already exists in the repo.
It's not a perfect solution, though.

I think in that case it would be better for the users to trim exif data themselves, otherwise they get a broken site if they lose their HDD...

cc @vojtechkral

**Comment by vojtechkral** [comment]
Hm. Filtering geo coords might be useful, but it seems like a fairly specialized thing...

**Comment by mkroman** [comment]
> I think in that case it would be better for the users to trim exif data themselves, *otherwise they get a broken site if they lose their HDD*...

Not necessarily. If the generated exifless-image uses a reproducible output path that gets checked in, it's possible to check if a file already exists at this generated output path before doing any work, so even if the original image is lost, the generated image should still exist and be valid.

On [another project][1], I'm already using *nix tools to accomplish this using a basic execution pipeline - I use `exiftran -i -a <file>` to rotate the image according to exif metadata and `exiv2 rm <file>` to remove exif metadata afterwards - this could be provided as an example solution and it would solve the problem, but it's very platform specific and I'd argue that there's practically no good reason for leaving these possibly very private details in *any* published images on the internet *in general* - unless the user explicitly requests it.

[1]: https://github.com/mkroman/meta-mail_reader/tree/c97ac01de4a67de7d0705cbe574a2364f09b0d23/lib/meta/postprocessors

**Comment by kellpossible** [comment]
To solve the rotation exif issue perhaps we could have an option to apply exif rotations to the image data. I've already run into some issues where images from smartphones don't display correctly in the browser due to the rotation being applied in the exif.

**Comment by mkroman** [comment]
I now have a different point of view to this problem; instead of messing around with `.gitignore`, pre-processing, repo hooks, or dynamic zola functions - why not make `zola` a complete command-line tool akin to how `cargo` or similar CLIs are used today?

I propose that `zola` either adds dynamic binary execution (i.e. `zola something` executes `zola-something` if found in `$PATH` akin to how `git` or `cargo` does it; details unclear) in order to expand its functionality - or - preferably, instead it embeds more CLI features that may be feature gated (such that each release includes 2 binaries: `zola-serve` (functionally equivalent to how it is now, used in production) and `zola` which includes `zola-serve` and all additional user CLI features))

On top of this CLI we could add a command à la `zola image add ~/Pictures/IMG_20191006_221543.jpg` which would, assuming we're in the project root directory, do the following:

* Open `~/Pictures/IMG_20191006_221543.jpg` 
* Parse image, optionally exclude metadata
* Write image to `static/images/IMG_20191006_221543.jpg`, possibly rotated, without metadata
* Print a helpful message similar to:  `In order to add this to your page, add the following: {{ thumbnailed_image(path='{relative_path_or_whatever}/IMG_20191006_221543.jpg')
  }}`

With this solution the end image is written to the (possibly public) project without leaking private metadata in a (relatively) reproducible manner that is CDN-friendly.

**Comment by mkroman** [comment]
A naive look into the currently related crates yields [img-parts](https://crates.io/crates/img-parts) which at first glance seems to offer an API to set a nullifying exif writer (i.e. `jpeg.set_exif(None)` in [this example](https://web.archive.org/save/https://docs.rs/img-parts/0.1.1/img_parts/) - mayhaps exif metadata could become a part of the [image](https://crates.io/crates/image/0.23.6) specs?

**Comment by Keats** [comment]
> I propose that zola either adds dynamic binary execution (i.e. zola something executes zola-something if found in $PATH akin to how git or cargo does it; details unclear) in order to expand its functionality - or - preferably, instead it embeds more CLI features that may be feature gated (such that each release includes 2 binaries: zola-serve (functionally equivalent to how it is now, used in production) and zola which includes zola-serve and all additional user CLI features))

That's kind of the opposite of the goal of the project though, "everything you need in one binary".

**Comment by mkroman** [comment]
> That's kind of the opposite of the goal of the project though, "everything you need in one binary".

That's true, and it was also just a recommendation to avoid a big change for a niche feature, and to keep the binary size small-ish, but it probably doesn't add up to that much anyway.

**Comment by bitinn** [comment]
I am late to this, I would like at least `resize_image` to have an option to clear EXIF data automatically, so we can implement a shortcode that take care of such work automatically.

Not sure how we should handle markdown images, perhaps a settings in front matter if possible?

**Comment by bitinn** [comment]
I thought more about this, if `resize_image` support output other than `processed_images`, then we could potentially make use of it to generate thumbnails that colocate with the content.

this is much more preferable for site with more articles, otherwise all these thumbnails are stuck inside a single `processed_images` folder.

**Comment by Keats** [comment]
I think in practice it shouldn't be done through the template engine at all for markdown images/colocation

**Comment by ebkalderon** [comment]
Any thoughts on implementing something vaguely akin to [`images.Filter`](https://gohugo.io/functions/images/filter/) from Hugo? This would enable users to [dynamically generate OpenGraph link preview images](https://runtimeterror.dev/dynamic-opengraph-images-with-hugo/#aaros-og-image-generation), for instance, without needing any external tools. Just spitballing below:

```jinja2
{%- set base_image = load_data(path="content/blog/my-article/og_base.png") -%}
{%- set bold_font = load_data(path="static/fonts/BoldFont.ttf") -%}
{%- set normal_font = load_data(path="static/fonts/NormalFont.ttf") -%}

{%- set og_image = base_image | image_filter(op="blur", sigma=8) | image_filter(op="text", text=config.title, font=bold_font, color="#FFFFFF", size=52, x=141, y=117) -%}
{%- set og_image = og_image | image_filter(op="text", text=page.title, font=normal_font, color="#FFFFFF", size=64, x=141, y=291) -%}

<meta property="og:image" content="{{ og_image.url | safe }}">
<meta property="og:image:width" content="{{ og_image.width }}">
<meta property="og:image:height" content="{{ og_image.height }}">
<meta property="og:image:type" content="{{ og_image.mime }}">
```

**Comment by Keats** [comment]
Yeah I don't see why we couldn't add that
